ligo/main.mligo

```diff
@@ -240,6 +240,75 @@ let set_position (s : storage) (p : set_position_param) : result =
       }
     )
 
+// Entrypoint that returns cumulative values at given range at the current moment
+// of time.
+let snapshot_cumulatives_inside (s, p : storage * sci_param) =
+    let sums = get_last_cumulatives s.cumulatives_buffer in
+    let cums_total =
+            { tick = sums.tick.sum
+            ; seconds = Tezos.now - epoch_time
+            ; seconds_per_liquidity = sums.lps.sum
+            } in
+
+    [@inline]
+    let eval_cums (above, index, cums_outside : bool * tick_index * cumulatives_data) =
+        // Formulas 6.22/23
+        if (s.cur_tick_index >= index) = above
+        then
+            { tick =
+                cums_total.tick - cums_outside.tick
+            ; seconds =
+                assert_nat(cums_total.seconds - cums_outside.seconds, internal_seconds_cumulative_err)
+            ; seconds_per_liquidity = {x128 =
+                assert_nat(cums_total.seconds_per_liquidity.x128 - cums_outside.seconds_per_liquidity.x128, internal_lps_cumulative_err)
+                }
+            }
+        else
+            cums_outside
+        in
+
+    // // TODO: looks like the caller must provide indices of already initialized ticks
+    // // But does it make sense? Should we accept tick witnesses explicitly here too?
+    let lower_tick = get_tick s.ticks p.lower_tick_index tick_not_exist_err in
+    let upper_tick = get_tick s.ticks p.upper_tick_index tick_not_exist_err in
```

Comment:
Looking better at it, it seems you are right.

What we use to initialize ticks is context-depended, so we need an actual `tick_state` to know those values.

Never mind my previous comment.

Reply:
As for the TODO, I think that requiring the caller to pick initialized ticks is the right way to go here.

This entrypoint is used to request data about specific ticks, so using witnesses here doesn't make sense to me.

I mean we are not adding new entries to the storage here and picking "close" ticks would (sort of silently) provide incorrect data, right?

IMO it makes more sense what you currently do: fail and have the caller retry in that case.

Reply:
Agreed :+1:  instead of using witnesses, we better expect the provided tick indices to refer to initialized ticks. In some cases, the caller can expect his indices to correspond to initialized ticks, e.g. if he created a position on them first.

> I mean we are not adding new entries to the storage here and picking "close" ticks would (sort of silently) provide incorrect data, right?

Hm, right, looks like we have no information to restore exact cumulative values at ticks in-between, In the best case we could try to interpolate their values, but without tick witness this is not possible anyway.